### ISSUE TITLE
What should we use as the eth-trie hash type?

### ISSUE THREAD

**Issue by carver** [carver/eth-trie.rs#30]
Two years ago, I made a choice to represent the hash as a keccak::H256. It's been long enough that the choice might not be the best one anymore (it was just a byte slice before, so it's still probably better than that)

When trying out v0.2.0, which includes this change, in trin, that seems like the biggest breaking change. So, what is the best option here?

One option would be to switch the type from the keccak one to the `ethereum_types` `H256`. That requires exposing ethereum_types as a proper dependency (not just a dev one). Also, it doesn't fully fix the issue:

```sh
error[E0277]: the trait bound `ethereum_types::H256: ssz::Decode` is not satisfied
  --> ethportal-api/src/types/consensus/body.rs:11:59
   |
11 | #[derive(Debug, PartialEq, Clone, Deserialize, Serialize, Decode, Encode)]
   |                                                           ^^^^^^ the trait `ssz::Decode` is not implemented for `ethereum_types::H256`
   |
   = help: the following other types implement trait `ssz::Decode`:
             (A, B)
             (A, B, C)
             (A, B, C, D)
             (A, B, C, D, E)
             (A, B, C, D, E, F)
             (A, B, C, D, E, F, G)
             (A, B, C, D, E, F, G, H)
             (A, B, C, D, E, F, G, H, I)
           and 98 others
   = note: this error originates in the derive macro `Decode` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `ethereum_types::H256: ssz::Decode` is not satisfied
   --> ethportal-api/src/types/consensus/body.rs:15:19
    |
15  |     pub graffiti: H256,
    |                   ^^^^ the trait `ssz::Decode` is not implemented for `ethereum_types::H256`
    |
    = help: the following other types implement trait `ssz::Decode`:
              (A, B)
              (A, B, C)
              (A, B, C, D)
              (A, B, C, D, E)
              (A, B, C, D, E, F)
              (A, B, C, D, E, F, G)
              (A, B, C, D, E, F, G, H)
              (A, B, C, D, E, F, G, H, I)
            and 98 others
note: required by a bound in `SszDecoderBuilder::<'a>::register_type`
   --> /home/carver/.cargo/registry/src/index.crates.io-6f17d22bba15001f/eth2_ssz-0.4.0/src/decode.rs:176:29
    |
176 |     pub fn register_type<T: Decode>(&mut self) -> Result<(), DecodeError> {
    |                             ^^^^^^ required by this bound in `SszDecoderBuilder::<'a>::register_type`
```

The most obvious options are:
- keep keccak:H256 as the exposed type, dependent projects just have to adjust (and maybe have the same encoding issues?)
- switch to ethereum_types:H256, and try to get the ssz lib to write encode/decode for this type
- switch it back to a bytes slice, and lose the guarantee of it being a 32-byte slice

I'll sleep on it. Open to discussion!
I guess we could try to get the ssz library to implement Encode and Decode for the H256 type

**Comment by carver** [comment]
@ogenev what do you think of these options?

**Comment by ogenev** [comment]
`eth2_ssz::Decode` and `Encode` are implemented in `ethereum_types::H256` for version `0.12.1`. 
It looks like this support is dropped for versions `0.13.1` and  `0.14.1`, `H256` is now re-exported from `primitive-types`: https://github.com/paritytech/parity-common/blob/f9cfc79fdd25b0bb0b96b7265055e82f932f299c/ethereum-types/src/hash.rs#L44 

Trin types are using `H256` from ethereum_types `0.12.1`: https://github.com/ethereum/trin/blob/cbede65de50e2eb74c8ccdbfe8bae39e5202ef9a/ethportal-api/Cargo.toml#L23 .

As a quick fix, switching to ethereum-types `0.12.1` should do the trick but in the long run, we may want to switch to a new SSZ library in Trin which seems to support  `H256` from ethereum types 0.14.1: see [lighthouse](https://github.com/sigp/lighthouse/blob/dfcb3363c757671eb19d5f8e519b4b94ac74677a/consensus/types/Cargo.toml#L28-L30).
